Implement CSS Properties & Values API in Servo.

jyc opened the pull request:
Here are some patches to implement the CSS Properties & Values API in Servo! I have a simple patch to implement it in Firefox via Stylo afterwards. I also have patches to implement animations for these in Stylo, but not in Servo.

One-line Gecko change to ServoBindings.toml required as a consequence of replacing custom_properties::CustomPropertiesMap with properties_and_values::CustomPropertiesMap in the ComputedValues.

Animations patches involve much more changes to Stylo than Servo, so I think those should be reviewed separately & through the Stylo process.

Some disabled tests:
- We don't implement `<transform-list>` serialization.
- We don't serialize colors according to spec & as Gecko does.
- We don't round floats correctly (see servo/rust-cssparser#173).
- We don't implement `<resolution>`.

It's not clear whether or not we actually need initialValue (see w3c/css-houdini-drafts#286).

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #17155 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18167)
<!-- Reviewable:end -->

jyc committed:
tests: Add tests for CSS Properties & Values API.

Some disabled tests:
- We don't implement <transform-list> serialization.
- We don't serialize colors according to spec & as Gecko does.
- We don't round floats correctly (see rust-cssparser/#173).
- We don't implement <resolution>.

It's not clear whether or not we actually need initialValue (see css-houdini-drafts/#286).

Part 13 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
style: Refine custom_properties::parse_name; disallow extra tokens.

Part 1 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
resources: Add a pref for CSS Properties & Values API.

The pref is layout.css.properties-and-values.enabled, and it currently
defaults to false. Future patches in this series will add code that uses
this pref.

Part 2 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
script: Add WebIDL and stubs for CSS.(un)registerProperty.

Part 3 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
style: Expose transform::SpecifiedValue's internals.

A later patch in this series will use this when to inspect transform
values for typed custom properties.

Part 4 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
script: Store typed custom property registrations on the document.

CSS Properties & Values API says the set of registered properties should
be associated with the document.

Part 7 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
layout_thread, style: Attach registered property set to the Stylist.

We store a pointer to the registered property set on the Stylist because
that seems to be the cleanest way to give access to it to all the
functions that might need to cascade or compute typed custom properties.

This also lets us use the Stylist's existing mechanism for determining
we need to rebuild styles: if it last saw the registered property set
with an older generation than the current generation, we should do
things again.

This is handled by the same path in layout_thread::lib as the device
being dirtied or the stylesheets changing.

Also update the comment on Stylist::rebuild, which appears not to have
been updated when stylesheets_changed was replaced with
origins_to_rebuild.

Part 9 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
style: Fix comments in style::properties.

"An CSS-wide keyword" should be "A CSS-wide keyword." Also, the
PropertyDeclaration::Custom variant must not contain a
DeclaredValueOwned::WithVariables variant: this is in fact already
asserted in style::custom_properties::cascade.

Part 10 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
script: Implement CSS.(un)registerProperty.

Part 12 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
style: Add/refactor types of custom_properties for Properties & Values.

Add a new ExtraData enum. We carry it along with a custom property's
specified value in order to compute it, if it ends up being registered
as being able to contain URLs through Properties & Values. When the
specified value comes from a declaration, we keep track of the
associated UrlExtraData in the Specified variant. However, specified
values will also (implemented by a later patch in this series) be able
to come from animations: in that case we are able to carry along a copy
of the computed value in the Precomputed variant, to save us from
recomputation and also to save us from having to re-resolve the URL.

Add a new BorrowedExtraData enum, which is used for
BorrowedSpecifiedValues. We use BorrowedSpecifiedValues when resolving
custom properties; in the case that we inherit an untyped custom
property, we create one from its computed value, which is a token
stream. In this case we don't have a UrlExtraData or (Properties &
Values) ComputedValue to compute with, and in fact we don't want to
perform any computation. The BorrowedExtraData enum thus has an
additional InheritedUntyped variant for that case.

Pull out the css, first_token_type, and last_token_type fields of
SpecifiedValue, BorrowedSpecifiedValue, and ComputedValue into a
TokenStream struct (ComptuedValue then becomes a type alias of
TokenStream). We'll use the TokenStream to store computed token stream
values for typed custom properties: it wouldn't make sense to use
SpecifiedValue because we have no need of the additional information
(and that would also lead to properties_and_values::ComputedValue
containing a SpecifiedValue which contains an ExtraData which may
contain a properties_and_values::ComputedValue, which the compiler does
not like).

The properties_and_values module is addded by a later patch in this
series, so we omit the Precomputed variant for now.

Part 5 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
style: Add a properties_and_values module for CSS Properties & Values.

Typed custom properties are exactly the same as regular custom
properties except when it comes to inheritance, computed values, and
animation, so this module is intended to be as independent from
custom_properties as possible (they do use types from each other,
though, where it would be clumsy to do otherwise).

This commit doesn't hook up the module to anything: that's handled by
later patches in this series.

Part 6 of a series of patches to implement the CSS Properties & Values API.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties_and_values.rs, components/style/properties/properties.mako.rs, components/style/lib.rs, components/style/stylist.rs, components/style/properties/gecko.mako.rs and 6 more
 * @canaltinova: components/style/properties_and_values.rs, components/style/properties/properties.mako.rs, components/style/lib.rs, components/style/stylist.rs, components/style/properties/gecko.mako.rs and 6 more
 * @kichjang: components/script/dom/document.rs, components/script/dom/bindings/trace.rs, components/script/dom/webidls/CSS.webidl, components/script/dom/css.rs, components/script/dom/webidls/PropertyDescriptorDict.webidl
 * @fitzgen: components/script/dom/document.rs, components/script/dom/bindings/trace.rs, components/script/dom/webidls/CSS.webidl, components/script/dom/css.rs, components/script/dom/webidls/PropertyDescriptorDict.webidl
 * @emilio: components/style/properties_and_values.rs, components/style/properties/properties.mako.rs, components/style/lib.rs, components/style/stylist.rs, components/style/properties/gecko.mako.rs and 6 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

emilio commented on components/style/stylist.rs:
```diff
@@ -98,6 +101,16 @@ pub struct Stylist {
+    /// The set of registered custom property associated with the document.
+    /// Should be initialized as soon as possible.
+    #[cfg_attr(feature = "servo", ignore_heap_size_of = "Arc")]
+    registered_property_set: Option<Arc<RwLock<RegisteredPropertySet>>>,
```
Why does this need to live in `Stylist`? It's only used to hand it out, right? If so, I suspect it should live in `SharedStyleContext` instead.

emilio replied:
The the generation, if any, should be either in `PerDocumentStyleDataImpl` in Gecko, or `LayoutThread` / `Document` in Servo.

emilio replied:
Or either just make it live on the stylist (since I see you require it from `cascade`), but forward commands to update it to layout from script. Would that work?

Also, that would prevent the `RwLock`.

emilio replied:
Hmm... I guess it won't work if we allow querying the custom properties from script, which is kinda annoying... :(

jyc replied:
Yep, can't make it live on the Stylist because the Stylist is not around for all the times someone might query from JS, unfortunately.

emilio replied:
Can we at least use a `SharedRwLock` instead? That'd avoid the atomic operations in `cascade`.

jyc replied:
Can try, but IIRC I ran into some issues. Might have been before I found the spec was changed to not require changing used declarations, though.

jyc replied:
Which SharedRwLock are you thinking of? Do you mean instead you'd just prefer if cascade did not do any locking at all wrt the RegisteredPropertySet? That would certainly make sense & is a good catch, thanks.

emilio replied:
FWIW what we do for stylesheets and similar is keeping two sets, one on layout and one on script (see the `Document::stylesheets` field). That makes the handling of them easier from style, so we may want to consider that, given Gecko wouldn't have this annoying problem of having to move the set across threads.

In any case, I haven't gone through it in detail, but I'd be very confused if we'd need to mutate this kind of global state each time we style an element, so I suspect we can at least get rid of the `RwLock`.

Also, I see no call to `set_registered_custom_properties` in Stylo, right?

jyc replied:
As noted, the Stylo CSS.registerProperty/unregisterProperty is in a separate patch. That goes through the same codepath (properties_and_values::register_property) as Servo's CSS.registerProperty/unregisterProperty.

Going to try using the SharedRwLock!

emilio commented on components/style/values/computed/length.rs:
```diff
@@ -194,21 +194,31 @@ impl From<LengthOrPercentageOrNone> for Option<CalcLengthOrPercentage> {
+        // 3. Otherwise, serialize as a calc() containing the summation, with
+        //    the units ordered ASCII case-insensitive alphabetically, the
+        //    number (if present) coming before all units, and the percentage
+        //    (if present) coming after all units.
```
Please undo this commit, we serialize percentage before intentionally per https://github.com/w3c/csswg-drafts/issues/1731.

jyc replied:
Hm, a comment would be nice. :P Will revert.

emilio replied:
Yep, I added it into the specified calc serialization, but not to the computed one, sorry! :)

emilio reviewed (commented):
Just a super-quick skim... I think I'd prefer no `RwLock`.

I've also noticed a bunch of other nits, but those are probably not a huge deal, and I'll comment with them later.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18143) made this pull request unmergeable. Please resolve the merge conflicts.

jyc committed:
layout_thread, style: Add guard for the registered property set.

Add a guard to StylesheetGuards for the RegisteredPropertySet.

Part 8 of a series of patches to implement the CSS Properties & Values API.

jyc committed:
style: Update cascade to take account of typed custom properties.

We have to do some additional work to compute typed custom properties
and to handle cyclical declarations involving font-size.

(Unfortunately there doesn't seem to be a really good way to split this
patch up without spending a ton of effort, and I don't think it'd make
the patch more clear. Sorry!)

Make custom_properties::OrderedMap::remove public, so that we can remove
cyclical custom property declarations from properties. We also need to
insert the initial value (possibly typed) if necessary and detect if
font size is cyclical and remove it in that case, so it makes sense to
handle them in the same place instead of custom_properties mutating
things behind the scenes and making it worry about typed custom
properties.

custom_properties::cascade takes two function arguments,
inherited_computed and handle_keyword. This allows us to maintain
separation of concerns wrt typed custom property registrations (also
makes it clearer who is getting data from where).

custom_properties::substitute_all and custom_properties::substitute_one
gains function arguments compute handle_invalid with the same
justification.

custom_properties::finish_cascade is replaced by
custom_properties::compute_ordering, which computes whether or not
font-size's declaration is cycles, the custom properties to compute before
early properties (including those upon which early properties depend),
and the set of properties that are cyclical.

Fix the comment on substitute, which does not type.

Have Gecko's ComputedValuesInner::get_custom_properties return a
properties_and_values::CustomPropertiesMap instead of a
custom_properties::CustomPropertiesMap. This stores typed computed
values as well.

Replace Servo's Stylist::custom_properties with
get_custom_properties, which returns a reference instead of an Arc,
because it doesn't make sense to clone the Arc when we never need it as
one. Also add get_parent_custom_properties, so a later patch in the
series can inherit typed custom property values.

Note that this also requires a change to ServoBindings.toml to account
for replacing custom_properties::CustomPropertiesMap with
properties_and_values::CustomPropertiesMap.

Part 11 of a series of patches to implement the CSS Properties & Values API.

jyc commented:
Changes:

- Lock the RegisteredPropertySet with a SharedRwLock.
- Update tests & code to account for initialValue being required when syntax is neither omitted nor "*".
- Remove fix to calc() serialization.
- Rebase.
- Fix commit message for the patch to update cascading; we removed StyleBuilder::custom_properties, not ComputedValues::custom_properties.

jyc commented:
Oops, looks like the changes for rebasing were left out. Pushing that.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18209) made this pull request unmergeable. Please resolve the merge conflicts.

jyc commented:
r? emilio

Hi! Manish suggested asking you to review--could you review this when you have the time? (if not, perhaps you have an idea of who else might review; perhaps the patches could be reviewed by separate people as well, like last time). I have some other patches that depend on it & those and this seem likely to continue to need rebasing.

emilio commented:
Sure, will take a look, sorry for not doing it before :)

jyc commented:
No problem, thanks!

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+        fn parse_custom_ident<'i, 't>(
+            _context: &ParserContext,
+            input: &mut Parser<'i, 't>,
+        ) -> Result<Ident, StyleTraitsParseError<'i>> {
```
You could use `CustomIdent::parse` here :)

jyc replied:
Nope, that is decidedly not a bug. See the definition of [<custom-ident>](https://drafts.csswg.org/css-values-4/#custom-idents). Adding a comment.

jyc replied:
https://github.com/w3c/css-houdini-drafts/issues/259#issuecomment-233049572

emilio replied:
Sure, I mean the opposite, right now there's nothing preventing them from being parsed, but `CustomIdent::parse` does handle it automatically.

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+        fn parse_custom_ident<'i, 't>(
+            _context: &ParserContext,
+            input: &mut Parser<'i, 't>,
+        ) -> Result<Ident, StyleTraitsParseError<'i>> {
```
I think this also prevents CSS wide keywords from being parsed, which I suspect is a bug too.

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+    CustomIdent,
+    /// Syntax to allow a specific identifier (matching the <custom-ident>
+    /// production, compared codepoint-wise).
+    SpecificIdent(Ident),
```
Oh, I guess this would match if I used a syntax like `foo bar`, then I'd have two `SpecificIdent`s, right?

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+                    // unfortunately.
+                    Type::SpecificIdent(ref x) => {
+                        $input.expect_ident_cloned().and_then(|y| {
+                            if idents_eq(&**x, &*y) {
```
This could use again `CustomIdent::parse($input)`, and then comparing `CustomIdent`s becomes cheap in Gecko too :)

jyc replied:
Can't seem to find a function with that name: https://doc.servo.org/style/values/struct.CustomIdent.html

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+
+impl Syntax {
+    /// Parse a syntax string given to `CSS.registerProperty`.
+    pub fn from_string(input: &str) -> Result<Syntax, ()> {
```
This needs a link to the spec too, I guess.

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+        // https://doc.servo.org/script/dom/bindings/str/struct.DOMString.html
+        // . This justifies iteration by |char|.
+
+        // Can identifiers in syntax strings contain escapes? No.
```
Ok, this means that what I pointed out about `Ident` is not a bug, I think. Still I'd prefer if we re-used the `CustomIdent` type, wdyt?

jyc replied:
Sounds good! Thanks for noticing that! :D

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+        // escapes, we need to check for escapes, which are only introduced by
+        // backslashes, which shouldn't show up anyhow.
+        // https://drafts.csswg.org/css-syntax-3/#escaping
+        let mut contains_backslash = false;
```
nit: This can be written like:

```rust
let contains_backslash = input.chars().any(|c| c == '\\');
```

Which is shorter and more idiomatic.

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+        /// Add a `Type` to the disjunction. It might turn out this is a list
+        /// term, in which case we'll modify the `Term` later.
+        fn push_type(syntax: &mut Option<Syntax>, t: Type) {
+            if let Some(Syntax::Disjunction(ref mut ts)) = *syntax {
```
nit: usually `match` is preferred to `if let ... { } else { unreachable!() }`, I believe.

emilio replied:
(This comment also applies to other occurrences in the file)

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+            if let Some(Syntax::Disjunction(_)) = *syntax {
+                // Good!
+            } else {
+                assert!(*syntax == None);
```
nit: Probably `debug_assert_eq!(*syntax, None)`. That provides a nicer error message.

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+        }
+
+        /// Handle the next token in the syntax string (including whitespace).
+        fn handle_token(syntax: &mut Option<Syntax>, state: State, token: &Token) -> Result<State, ()> {
```
I _suspect_ that this can be cleaner having some sort of `SyntaxParser` object that contains the `Option<Syntax>`, and presumably the `State`, and implementing methods on it, instead of a lot of free functions. In rust in-out parameters aren't very common.

I suspect it's not a big deal, and can be cleaned up later, but it'd be really nice to see it. Your call.

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+                },
+                (State::Start { after_bar: false }, &Token::Delim('*')) => {
+                    // If we see a '*', that should be it (modulo whitespace).
+                    if syntax != &None {
```
`if syntax.is_some()` instead.

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+                (State::AfterOpenAngle, &Token::Ident(ref id)) => {
+                    // We should be in something like '<length>' here.
+                    // https://drafts.css-houdini.org/css-properties-values-api/#supported-syntax-strings
+                    push_type(syntax, match &**id {
```
Should this be ascii-case-sensitive? or insensitive? (need to go through the spec as a whole, it's pretty late here so I'm doing a first pass to get the overview of how this looks into my mind for tomorrow :P)

jyc replied:
ASCII-case-sensitive.

jyc replied:
Well, rather, codepoint matched (not familiar with ASCII-case-sensitive). https://github.com/w3c/css-houdini-drafts/issues/259#issuecomment-233049572

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+                },
+                Err(_) => return Err(()),
+                Ok(token) => {
+                    match handle_token(&mut syntax, state, token) {
```
nit: You can write this like:

```rust
state = handle_token(&mut syntax, state, token)?;
```

Not sure how you feel about the extra magic.

emilio commented on components/style/custom_properties.rs:
```diff
@@ -251,17 +381,26 @@ impl ComputedValue {
+    /// Returns whether or not this specified value contains any variable
+    /// references.
+    pub fn has_references(&self) -> bool {
+        !self.references.as_ref().map(|x| x.is_empty()).unwrap_or(true)
```
nit: `map_or`

jyc replied:
That is awesome! Thanks

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+/// A custom property syntax string.
+#[derive(Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum Syntax {
```
I think this could be on its own module, along with the parser, what do you think?

Also, this could land before, with a few unit tests, and such.

Maybe it's time for `custom_properties` to be a folder on its own... :)

jyc replied:
Hm, I am not sure I agree. This module is the same size as `custom_properties`. Also, I am not too clear on the value of unit tests that can only be run with Servo when we have web-platform tests that should be better maintained because they can be run with all web browsers.

emilio replied:
Sure, I meant as a stop gap, until all the extra support for exposing it lands.

emilio replied:
We've done that before for some CSS properties and such.

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+    /// Returns whether or not this specified value is computationally
+    /// independent.
+    /// See the comment on SpecifiedValueItem::is_computationally_independent.
+    pub fn is_computationally_independent(&self) -> bool {
```
Yeah, this indeed sounds a lot like a trait that we could `derive` :)

jyc replied:
Certainly!

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+/// If `list` and `x` are both `Ok(..)`, append the contents of `x` to `list`.
+/// To be used as an argument to fold to convert a sequence of `Result<T, E>`s
+/// to a `Result<Vec<T>, E>` (pulling the `Result` outside).
+fn all<T, E>(list: Result<Vec<T>, E>, x: Result<T, E>) -> Result<Vec<T>, E> {
```
I don't see this function called at all?

jyc replied:
Not called, but used as the comment describes.

jyc replied:
And also used in the animation-related code (split into a separate PR).

emilio commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+    }
+}
+
+impl DerefMut for CustomPropertiesMap {
```
This looks like a bit of a hack to avoid changing more code :)

It's fine I suppose, but...

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1200,8 +1200,9 @@ impl LayoutThread {
         let ua_or_user_guard = ua_stylesheets.shared_lock.read();
         let guards = StylesheetGuards {
-            author: &author_guard,
+            author: &document_guard,
```
I think this doesn't need another guard. `author` is the origin of the stylesheets added by the author, so we can just use it instead of renaming it, can't we?

jyc replied:
The `document_guard` comes from the `style_shared_lock` on the document, and is used for both stylesheets added by the author as well as (now) the registered property set. If you think renaming the field on `StylesheetGuards` from `author` to `document` would make more sense (but that seems to be crossing the abstraction barrier) then I can do that.

emilio replied:
I think I'd prefer it, yeah. Author sheets are also document sheets, so it doesn't sound really bad.

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1137,6 +1137,18 @@ impl LayoutThread {
+        // regarding device changes below.
+        self.stylist.set_registered_property_set(document.registered_property_set());
+        if self.stylist.registered_property_set_updated(&document_guard) {
+            self.stylist.force_stylesheet_origins_dirty(OriginSet::all());
```
You only need to pass `Origin::Author.into()` here.

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1137,6 +1137,18 @@ impl LayoutThread {
+        // Set the registered property set (if it isn't already set), and mark
+        // stylesheet origins dirty if they've changed, analogously to the check
+        // regarding device changes below.
+        self.stylist.set_registered_property_set(document.registered_property_set());
```
I still think this API is not great... :/

emilio commented on components/style/stylist.rs:
```diff
@@ -458,6 +471,38 @@ impl Stylist {
+        &mut self,
+        registered_property_set: Arc<Locked<RegisteredPropertySet>>
+    ) {
+        debug_assert!(if let Some(ref old) = self.registered_property_set {
```
nit: Again, `as_ref().map_or(true, ...)` instead of `if let (..) { } else { .. }`.

jyc replied:
Nice catch!

emilio commented on components/style/stylist.rs:
```diff
@@ -458,6 +471,38 @@ impl Stylist {
+            .expect("set_registered_property_set must be called.")
+            .read_with(guard);
+
+        self.last_used_registered_property_set_generation != Some(registered_property_set.generation())
```
There's nothing that sets `self.last_used_registered_property_set_generation`, right? That looks like a bug to me.

jyc replied:
Fell out while rebasing, it seems. Seems we can remove this per your comment below.

emilio commented on components/style/stylist.rs:
```diff
@@ -458,6 +471,38 @@ impl Stylist {
+        let registered_property_set =
+            self.registered_property_set.as_ref()
+            .expect("set_registered_property_set must be called.")
+            .read_with(guard);
```
nit: This could do `self.registered_property_set().read_with(guard)` instead.

emilio commented on components/style/animation.rs:
```diff
@@ -494,6 +494,8 @@ fn compute_style_for_animation_step(context: &SharedStyleContext,
             // This currently ignores visited styles, which seems acceptable,
             // as existing browsers don't appear to animate visited styles.
+            let registered_property_set =
+                context.stylist.registered_property_set().read_with(context.guards.registered_property_set);
```
Ugh, I didn't remember we had `context.stylist` down here. Looks fine I guess, though it seems prone to abuse.

So far the only two callers are these too, so it's fine.

emilio commented on components/style/custom_properties.rs:
```diff
@@ -640,109 +672,252 @@ pub fn cascade<'a>(custom_properties: &mut Option<OrderedMap<&'a Name, BorrowedS
+/// Replace `var()` functions for all custom properties.
+pub fn substitute_all<C, H>(
+    specified_values_map: &OrderedMap<&Name, BorrowedSpecifiedValue>,
+    to_substitute: &Option<HashSet<Name>>,
```
nit: Instead of this, looks like it could take a `Option<&HashSet>`.

emilio commented on components/style/stylist.rs:
```diff
@@ -1504,7 +1517,12 @@ impl Stylist {
                             CascadeFlags::empty(),
-                            self.quirks_mode)
+                            self.quirks_mode,
+                            self.registered_property_set
```
nit: can use `self.registered_property_set()`, here and above.

jyc replied:
Great catch!

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -3085,9 +3090,338 @@ pub fn cascade(
+/// any properties and there are no uninherited custom properties).
+/// Returns the custom properties' specified nad computed values, as well as
+/// whether or not the declaration for font-size is involved in a cycle.
+fn compute_early_custom_properties<'a, 'b: 'a, I>(
```
If this doesn't need mako code, I'd really appreciate to have it outside of this file. I'm trying to remove it eventually.

jyc replied:
I think this should be in the same file that `cascade` is in as a separate function, in order to maintain modularity. If the functionality of `cascade` independent of mako code is eventually moved somewhere else, it should move with it, I think.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -3085,9 +3090,338 @@ pub fn cascade(
+/// all custom properties. Computed values are added to `computed`.
+///
+/// This function is called twice: before and after we compute early properties.
+fn substitute_all(
```
Ditto re. moving outside of mako files.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -3085,9 +3090,338 @@ pub fn cascade(
+    url_data: &UrlExtraData,
+    name: &::custom_properties::Name,
+    value: ::custom_properties::ComputedValue,
+) -> Result<(::custom_properties::ComputedValue,
```
Seems like all this triples could benefit from being a struct :)

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -3085,9 +3090,338 @@ pub fn cascade(
         font_metrics_provider,
         flags,
         quirks_mode,
+        registered_property_set.read_with(guards.registered_property_set),
```
I'd require just a `&RegisteredPropertySet`, and move the `read_with` to the caller when possible here, but no big deal.

jyc replied:
Did for consistency with other arguments to `cascade`.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -3085,9 +3090,338 @@ pub fn cascade(
+            // b) has an initial value,
+            // then we should set it to its initial value.
+            let mut computed = {
+                let has_value = specified.iter().map(|(name, _)| name.clone()).collect();
```
This doesn't seem particularly fast... But I guess

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -3085,9 +3090,338 @@ pub fn cascade(
+            // There were no inherited or specified values.
+            // If there are any properties with initial values, we should set
+            // them.
+            (Some(Arc::new(registered_property_set.initial_values_except(context, None))), false)
```
So we're creating this and calling `initial_values_except` for every element, even if we don't use custom properties?

That seems quite suboptimal, not only for the unnecessary allocation specially since the time `initial_values_except` takes grows with the number of registered properties :/

Also, no branch seems to return `None` for `computed_early`, is it intentional? If not, something needs to get fixed. If it is, I don't see why it should return an `Option`.

jyc replied:
> So we're creating this and calling initial_values_except for every element, even if we don't use custom properties?
> That seems quite suboptimal, not only for the unnecessary allocation specially since the time initial_values_except takes grows with the number of registered properties :/

Well, creating the Arc is expensive. But the cost of `initial_values_except` would not apply in the case you mentioned except on the root, as either we are inheriting precisely these values, or in fact there are no registered properties whatsoever.

I will fix it so that if there are no registered properties we return None as we would have in the past. In the future, and in line with https://github.com/w3c/css-houdini-drafts/issues/286, it seems we should really be not storing computed initial values here at all (so that we can be completely analogous with regular properties); will have to make sure to do in a nice way so that we don't violate the abstraction barrier between substition and registrations.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -3194,12 +3534,39 @@ where
             let mut font_size = None;
             let mut font_family = None;
         % endif
+        %if category_to_cascade_now == "other":
```
nit: Space after `% `.

emilio commented on components/script/dom/css.rs:
```diff
@@ -3,19 +3,42 @@
+    match result {
+        PropertyRegistrationResult::Ok => {
+            // Should lead to the RestyleHint::restyle_subtree() being inserted
+            // eventually in handle_reflow due to checks in Stylist.
```
No need to do this. Afaict you could just mark the document's `stylesheet` list as dirty from here, and everything else should follow. You'd even avoid needing the `registered_property_set_generation`!

jyc replied:
Nice! Would be cool if this kind of thing were documented.

jyc replied:
Neat, this seems to be another consequence of the spec change.

emilio reviewed (commented):
This is just a first pass, since it's quite late.

I think overall it looks reasonable. I'm slightly concerned about the perf impact even when custom properties aren't used.

When they're used, it seems like there's a fair amount of `HashSet` stuff going around, and I'd probably need to review the details to ensure we're not creating a lot of `HashSet`s unconditionally.

In general, I think this could've benefited from a bit less macro-patch, and trying to land stuff more incrementally.

For example, seems to me that all the syntax parsing code could've been an standalone PR. Maybe stubbing the methods could've been another one, and so on.

That'd make this way easier to review.

I think @SimonSapin and @heycam are familiar with the custom properties spec, so if they'd take a look I'd also appreciate it.

I think there are a few bits that look somewhat fishy, specially the generation stuff, which would force us to always do a stylist rebuild all the time.

It's possible that I've missed some stuff, but I have a fair amount of comments, I think :P

Also, thanks a lot for working on this! It's quite awesome. Let's ensure we get the details right :)

emilio commented:
One last bit. `Resolution` is only used for media query stuff, so it lives in `gecko/media_queries.rs`. A PR moving it to `values` would be nice.

Also, I'd really prefer everything to be in smaller modules, when possible. They're much nicer to read. Specially I'd love if it'd be avoidable to put more stuff in our mako files, which are already pretty bloated, and I'd like for them to eventually disappear.

jyc commented:
Sorry for the delay–ran into some confusion moving in this year. Should be settled down now! Going through review comments.

>  Also, I'd really prefer everything to be in smaller modules, when possible. They're much nicer to read.

Hm, the properties_and_values module is approximately the same size as the custom_properties module. I've already separated them in order to keep the functionality as distinct as possible; it's not clear to me how making the implementation of, say, typed custom property animation in a separate module from the definition of what types typed custom properties can have includes readability. If you have concrete suggestions about what to split, I'm happy to try them, though.

jyc commented on components/style/custom_properties.rs:
```diff
@@ -28,10 +28,17 @@ pub type Name = Atom;
-        Ok(&s[2..])
-    } else {
-        Err(())
+    let mut input = ParserInput::new(s);
```
Thanks! Will update the commit message. `parse_name` doesn't implement the spec that it cites:

> A custom property is any property whose name starts with two dashes (U+002D HYPHEN-MINUS), like --foo. The <custom-property-name> production corresponds to this: it’s defined as any valid identifier that starts with two dashes.

`s` needs not to not just start with two dashes, but to also be an identifier.

SimonSapin replied:
This function is intended to be with the value of `Token::Identifier(_)`, after CSS backslash-escaping has already been unescaped, or with a string coming from a CSSOM API where the property name is already a separate argument and does not need to using CSS syntax to delimit an identifier. Using a CSS parser again here is incorrect.

https://drafts.css-houdini.org/css-properties-values-api/#register-a-custom-property does suggest using a CSS parser but that’s wrong. I’ve filed https://github.com/w3c/css-houdini-drafts/issues/465.

jyc commented on components/style/custom_properties.rs:
```diff
@@ -42,20 +45,106 @@ pub fn parse_name(s: &str) -> Result<&str, ()> {
+pub enum ExtraData {
+    /// The specified value comes from a declaration (whence we get
+    /// the UrlExtraData).
+    Specified(UrlExtraData),
```
Thanks! Will add link to https://github.com/w3c/css-houdini-drafts/issues/393#issuecomment-294706386 . Other variants are added in subsequent patches.

jyc commented on components/style/custom_properties.rs:
```diff
@@ -42,20 +45,106 @@ pub fn parse_name(s: &str) -> Result<&str, ()> {
+    pub last_token_type: TokenSerializationType,
+}
+
+impl Default for TokenStream {
```
Thanks for catching this–this is used by the PR (not yet made, waiting for this one to land) implementing animations. Will move it over to the PR for that.

jyc commented on components/style/custom_properties.rs:
```diff
@@ -10,14 +10,17 @@ use Atom;
 use cssparser::{Delimiter, Parser, ParserInput, SourcePosition, Token, TokenSerializationType};
 use parser::ParserContext;
 use properties::{CSSWideKeyword, DeclaredValue};
+use properties_and_values;
```
Hm, the module is added in the next commit; not sure how this line ended up in the previous one. Will fix, thanks!

jyc commented on components/style/custom_properties.rs:
```diff
@@ -42,20 +45,106 @@ pub fn parse_name(s: &str) -> Result<&str, ()> {
+impl Default for TokenStream {
+    fn default() -> Self {
+        TokenStream {
+            css: "".to_owned(),
```
Sounds good!

jyc commented on components/script/dom/webidls/PropertyDescriptorDict.webidl:
```diff
@@ -0,0 +1,17 @@
+ */
+
+// Renamed from PropertyDescriptor to avoid conflicting with a JS class of the
+// same name.
```
See https://bugzilla.mozilla.org/show_bug.cgi?id=1273706#c325.

jyc replied:
(Though actually, the reason might no longer apply. Going to try!)

jyc replied:
No, unfortunately we still have that problem. :(

```
[jyc@jc-mbp-osx servo] (git pv/HEAD tests: Add tests …)
→ ./mach build
   Compiling script v0.0.1 (file:///Users/jyc/projects/servo/components/script)
   Compiling style v0.0.1 (file:///Users/jyc/projects/servo/components/style)
   Compiling gfx v0.0.1 (file:///Users/jyc/projects/servo/components/gfx)
   Compiling metrics v0.0.1 (file:///Users/jyc/projects/servo/components/metrics)
   Compiling layout_traits v0.0.1 (file:///Users/jyc/projects/servo/components/layout_traits)
   Compiling script_layout_interface v0.0.1 (file:///Users/jyc/projects/servo/components/script_layout_interface)
   Compiling constellation v0.0.1 (file:///Users/jyc/projects/servo/components/constellation)
   Compiling layout v0.0.1 (file:///Users/jyc/projects/servo/components/layout)
error[E0252]: the name `PropertyDescriptor` is defined multiple times
   --> /Users/jyc/projects/servo/target/debug/build/script-9d08ccc9e36b2d2c/out/Bindings/CSSBinding.rs:443:5
    |
251 | use dom::bindings::codegen::Bindings::PropertyDescriptorBinding::PropertyDescriptor;
    |     ------------------------------------------------------------------------------- previous import of the type `PropertyDescriptor` here
...
443 | use js::jsapi::PropertyDescriptor;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `PropertyDescriptor` reimported here
    |
    = note: `PropertyDescriptor` must be defined only once in the type namespace of this module

error[E0255]: the name `PropertyDescriptor` is defined multiple times
   --> /Users/jyc/projects/servo/target/debug/build/script-9d08ccc9e36b2d2c/out/Bindings/PropertyDescriptorBinding.rs:239:1
    |
204 |   use js::jsapi::PropertyDescriptor;
    |       ----------------------------- previous import of the type `PropertyDescriptor` here
...
239 | / pub struct PropertyDescriptor {
240 | |     pub inherits: bool,
241 | |     pub initialValue: Option<DOMString>,
242 | |     pub name: DOMString,
243 | |     pub syntax: DOMString,
244 | | }
    | |_^ `PropertyDescriptor` redefined here
    |
    = note: `PropertyDescriptor` must be defined only once in the type namespace of this module

error: aborting due to 2 previous errors

error: Could not compile `script`.

To learn more, run the command again with --verbose.
Build FAILED in 0:01:39
```

jyc commented on components/style/lib.rs:
```diff
@@ -119,6 +119,7 @@ pub mod matching;
 pub mod media_queries;
 pub mod parallel;
 pub mod parser;
+pub mod properties_and_values;
```
Yep! Thanks, the ordering was messed up (they were supposed to build independently). Fixing!

jyc commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+/// resolve URLs when running in Stylo yet.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct ComputedUrl(pub String);
```
All of these variants duplicate names, to do so otherwise would make things very ugly (`properties_and_values::SpecifiedValue::PropertiesAndValuesSpecifiedAngle`?)

jyc commented on components/style/properties_and_values.rs:
```diff
@@ -0,0 +1,1167 @@
+    CustomIdent,
+    /// Syntax to allow a specific identifier (matching the <custom-ident>
+    /// production, compared codepoint-wise).
+    SpecificIdent(Ident),
```
You can't use a syntax like `foo bar`, although you could do something like `foo\20bar` or `foo | bar`. This regards the

```
Any sequence consisting of a name-start code point, followed by zero or more name
code points, which matches the <custom-ident> production

    That identifier
```

case. Thanks!

jdm commented:
Closing due to inactivity. Tracking this work in https://github.com/servo/servo/issues/17155.
